enhancements/network/ingress-node/ingress-node-firewall-with-ebpf.md

```diff
@@ -0,0 +1,284 @@
+## Summary
+
+OCP customers would like to secure OCP's nodes from external attacks by configuring ingress firewall policies.
+eBPF provides a very flexible filtering mechanism to allow early detection and filtering.
+eXpress Data Path (XDP) is a feature in the Linux kernel which allows users to execute a user-supplied eBPF program when a packet is received on a network interface (NIC). 
+This program is executed in the NIC’s device driver, right after the interrupt processing and before the network stack. 
+The goal of XDP is to offer comparable performance to kernel bypass solutions while working with the existing kernel networking stack.
+A new cluster-scoped object `IngressNodeFirewall` will be used to configure node ingress firewall match rules and per rule statistics will be avaliable.
+Admin can provision multiple instances of this object while ensuring certain execution order.
+For the best performance XDP programs need to attach to interfaces using the following models:
+- Native XDP \
+  The XDP program is loaded by the network card driver as part of its initial receive path. This also requires support from the network card driver.
+- Offloaded XDP \
+  The XDP program loads directly on the NIC, and executes without using the CPU. This also requires support from the network interface device.
+
+## Motivation
+
+Customers would like to secure OCP cluster from external attacks.
+
+### Goals
+
+- Flexible match filter rules for customers to use with a CRD validation to detect and reject invalid configuration.
+- Provide a mechanism for customers to select which nodes via `nodeSelector`.
+- Provide a mechanism to define attachment interfaces for the node ingress firewall objects.  
+- Ingress node firewall is deployed on a fully operational OCP cluster (i.e. Ingress node firewall is deployed as a day 2 operation and is not enabled by default at the cluster install time).
+- fail safe in case of user provisioned invalid rules which can lockup cluster operation.
+
+### Non-Goals
+- Stateful support
+- Flood Attacks
+
+## Proposal
+
+- customers will label nodes where ingress node policy will be deployed on.
+- customers will deploy daemonSet controller on all the selected nodes by applying `IngressNodeFirewallConfig` object. 
+- customers will prepare `IngressNodeFirewall` objects with all required firewall rules and interfaces the ebpf program will be attached to.
+- IngressNodeFW controller will process `IngressNodeFirewall` objects, verification webhook will make sure configuration doesn't contradict with fail safe rules.
+- Ingress node firewall daemon will watch ingress node firewall objects and create `nodeState` object which contains all rules objects that is applied to specific node then load eBPF map table and attach eBPF kernel hook loaded as `BPF_PROG_TYPE_XDP` to the list of interfaces configured in the object.
+- in order to fail safe and avoid cluster lockup, various pinholes will be open in the firewall to allow basic services.
```

Comment:
(I don't want to reply inside a resolved comment since it will get lost, but @astoycos had asked "Are these explicitly defined by the cluster admin? Or is it done internally in OCP whenever a new object is created?")

There was some debate about whether/how much we should have these, vs being 100% pro-shooting-yourself-in-the-foot.

Maybe a future update to this could have the failsafes also defined by a CRD (with an admission controller ensuring that it always contains a few specific rules, like allowing all the masters to talk to each other on the etcd and apiserver ports).

Then to screw things up you'd have to first accidentally remove the failsafes, and then accidentally create rules that blocked ssh, etc

Reply:
we could look at making failsafe a CRD for future release but the prog if that goes wrong we will go back to square 1 and lockup the cluster honestly don't see much benifit atm making them CRD